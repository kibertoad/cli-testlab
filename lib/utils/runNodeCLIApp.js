const path = require('path');

/**
 *
 * @param {string} pathToApp
 * @param {string[]} [args] - cmdline args that will be passed to application
 */
function runNodeCLIApp(pathToApp, args = []) {
  process.argv.push(...args);
  require(path.resolve(pathToApp));
}

module.exports = {
  runNodeCLIApp
};
