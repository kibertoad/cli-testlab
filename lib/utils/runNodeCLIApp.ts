import path from 'path'

/**
 * Load and execute given JS script file. Uses Node.js require instead of native CLI execution. This allows to use debugger on
 * script under test, but does not constitute a full e2e test. Recommended to use for debugging purpose, but not for tests that
 * will be executed in CI
 * @param {string} pathToApp - path to JS file
 * @param {string[]} [args] - cmdline args that will be passed to application
 * @param resetArgv - remove preexisting argv values (useful when used with Jest to avoid argv pollution)
 */
export function runNodeCLIApp(
  pathToApp: string,
  args: any[] = [],
  { resetArgv = false }: { resetArgv?: boolean } = {},
) {
  if (resetArgv) {
    process.argv = process.argv.slice(0, 2)
  }
  process.argv.push(...args)
  require(path.resolve(pathToApp))
}
