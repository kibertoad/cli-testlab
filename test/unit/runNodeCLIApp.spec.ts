import { runNodeCLIApp } from '../../lib/utils/runNodeCLIApp'
import path from 'path'

const pathToApp = path.normalize(__dirname + '/../util/apps/basic.cli.app.js')

describe('runNodeCLIApp', () => {
  it('executes with an error', async () => {
    expect(() => {
      runNodeCLIApp(pathToApp, 'error Kaboom'.split(' '), { resetArgv: true })
    }).toThrowErrorMatchingSnapshot()
  })

  it('executes without an error', async () => {
    expect(() => {
      runNodeCLIApp(pathToApp, 'message Okay'.split(' '), { resetArgv: true })
    }).not.toThrowError()
  })
})
