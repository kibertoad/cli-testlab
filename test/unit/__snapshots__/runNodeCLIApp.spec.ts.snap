// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runNodeCLIApp executes with an error 1`] = `"Kaboom"`;
